app/models/course_user_datum.rb

```diff
@@ -249,6 +249,88 @@ def ggl_cache_key
     "ggl/dua-#{dua}/u-#{self.id}"
   end
 
+  # This method call is used specifically for the purpose of callback style update to watchlist
+  # Need to archive old instances and also add new instances
+  def update_cud_gdu_watchlist_instances
+    # At this point, all relevant previously cached grace day usage information should be invalidated
+    # Calls to calculate grace day usage should be from scratch
+    
+    # Ignore if this CUD is an instructor or CA
+    return unless self.student?
+    
+    # Get current grace day condition
+    current_conditions = RiskCondition.get_current_for_course(self.course.name)
+    return if current_conditions.count == 0
+    
+    # Check for whether there exists one of type :grace_day_usage
+    grace_day_usage_condition = current_conditions.select { |c| c.condition_type.to_sym == :grace_day_usage }
+    return if grace_day_usage_condition.count != 1
+    grace_day_usage_condition = grace_day_usage_condition[0]
+    
+    # Get :grace_day_threshold and :date
+    condition_id = grace_day_usage_condition.id
+    grace_day_threshold = grace_day_usage_condition.parameters[:grace_day_threshold].to_i
+    date = grace_day_usage_condition.parameters[:date]
+    
+    # Archive old ones if there exist some
+    old_instances = self.watchlist_instances.where(risk_condition_id: condition_id)
+    
+    # Do the usual stuff as in refresh watchlist instances
+    # Create a new instance if criteria are fit
+    new_instance = self.new_gdu_watchlist_instance(grace_day_threshold, date, condition_id)
+    
+    # Make sure to use transaction
+    ActiveRecord::Base.transaction do
+      old_instances.each do |inst|
+        inst.archive_watchlist_instance
+      end
+
+      unless new_instance.nil?
+        if not new_instance.save
+          raise "Fail to create new watchlist instance for CUD #{self.id} in course #{self.course.name} with violation info #{new_instance.violation_info}"
+        end
+      end
+    end
+  end
+
+  # This returns an unsaved watchlist instance if current course user datum qualifies for
+  # being put onto the watchlist. The caller of this method needs to save it on their own if necessary
+  # This can be called from:
+  # - Watchlist's bulk refreshing
+  # - Callback style update within course user datum (see above)
+  # - Callback style update on a course level (see course.rb)
+  def new_gdu_watchlist_instance(grace_day_threshold, date, condition_id)
```

Comment:
I feel that this should be in the watchlist instance as well, because this adds a new watchlist instance, and not modifies course_user_datum